src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs

```diff
@@ -0,0 +1,176 @@
+                    {
+                        blocksToGoBack--;
+                    }
+                }
+
+                currBlockNumber--;
+            }
+
+            return gasPrices;
+        }
+        
+        private SortedSet<UInt256> eth_initializeValues(Block headBlock, Block genesisBlock, out long blocksToGoBack, out int percentile,
+            out long threshold, out long currBlockNumber, out long genBlockNumber, out UInt256? gasPriceLatest)
+        {
+            Comparer<UInt256> comparer = Comparer<UInt256>.Create(((a, b) =>
+            {
+                int res = a.CompareTo(b);
+                return res == 0 ? 1 : res;
+            }));
+            SortedSet<UInt256> gasPrices = new(comparer); //allows duplicates
+            blocksToGoBack = 5;
+            percentile = 20;
+            threshold = blocksToGoBack * 2;
+            currBlockNumber = headBlock!.Number;
+            genBlockNumber = genesisBlock!.Number;
+            gasPriceLatest = eth_latestGasPrice(headBlock!.Number, genesisBlock!.Number);
+            return gasPrices;
+        }
+        
+        private static bool eth_initalChecks(Block? headBlock, Block? genesisBlock, out ResultWrapper<UInt256?> resultWrapper)
+        {
+            if (headBlock == null || genesisBlock == null)
+            {
+                {
+                    resultWrapper = ResultWrapper<UInt256?>.Fail("The head block or genesis block had a null value.");
+                    return true;
+                }
+            }
+            
+            if (_lastPrice != null && _lastHeadBlock != null)
+            {
+                if (headBlock.Hash == _lastHeadBlock.Hash)
+                {
+                    {
+                        resultWrapper = ResultWrapper<UInt256?>.Success(_lastPrice);
+                        #if DEBUG
+                            resultWrapper.ErrorCode = -1;
+                        #endif
+                        return true;
+                    }
+                }
+            }
+
+            resultWrapper = ResultWrapper<UInt256?>.Success(UInt256.Zero);
+            return false;
+        }
+        
+        public ResultWrapper<UInt256?> eth_gasPrice(UInt256? ignoreUnder = null)
+        {
+            Block? headBlock = _blockFinder.FindHeadBlock();
+            Block? genesisBlock = _blockFinder.FindGenesisBlock();
+            if (eth_initalChecks(headBlock, genesisBlock, out ResultWrapper<UInt256?> resultWrapper))
+            {
+                return resultWrapper;
+            }
+
+            SortedSet<UInt256> gasPrices = eth_initializeValues(headBlock, genesisBlock, out long blocksToGoBack,
+                out int percentile, out long threshold, out long currBlockNumber, out long genBlockNumber, out UInt256? gasPriceLatest);
+            if (gasPriceLatest == null)
+            {
+                return ResultWrapper<UInt256?>.Fail("gasPriceLatest was not set properly.");
```

Comment:
the error message here looks like a message for developer but it is sent out to the user